### PR TITLE
[vcpkg] Add environment variable VCPKG_DEFAULT_BINARY_CACHE

### PR DESCRIPTION
**Pull request by Neumann-A**
Add an environment variable to reroute the binarychaching feature away from the users local data folder.

Somebody missed to add a user bailout from the default behavior ;)

Also:
Why the heck is that thing return `const ExpectedS<fs::path>&` of a static than using a lambda which returns `ExpectedS<fs::path>` which than again is in a function returning `const ExpectedS<fs::path>&`. Can we not agree to just return `const ExpectedS<fs::path> ` for RVO instead of a function local static variable which gets turned into another function local static variable?

Docs? Not yet, somebody has to add docs for binary caching anyway in the future. I'll leave it to him/her;)

### CODE REVIEW

**Comment by ras0219** [comment]
From `vcpkg help binarycaching`:
```
> ./vcpkg help binarycaching
Vcpkg can cache compiled packages to accelerate restoration on a single machine or across the
network. This functionality is currently enabled by default and can be disabled by either passing
`--no-binarycaching` to every vcpkg command line or setting the environment variable
`VCPKG_FEATURE_FLAGS` to `-binarycaching`.

Once caching is enabled, it can be further configured by either passing `--binarysource=<source>`
options to every command line or setting the `VCPKG_BINARY_SOURCES` environment variable to a set of
sources (Ex: "<source>;<source>;..."). Command line sources are interpreted after environment
sources.

Valid source strings:
  clear                           Removes all previous sources
  default[,<rw>]                  Adds the default file-based location.
  files,<path>[,<rw>]             Adds a custom file-based location.
  nuget,<uri>[,<rw>]              Adds a NuGet-based source; equivalent to the `-Source` parameter
                                  of the NuGet CLI.
  nugetconfig,<path>[,<rw>]       Adds a NuGet-config-file-based source; equivalent to the `-Config`
                                  parameter of the NuGet CLI. This config should specify
                                  `defaultPushSource` for uploads.
  interactive                     Enables interactive credential management for some source types

The `<rw>` optional parameter for certain strings controls whether they will be consulted for
downloading binaries and whether on-demand builds will be uploaded to that remote. It can be
specified as 'read', 'write', or 'readwrite'.
```

So, the existing way to do exactly this is to define `VCPKG_BINARY_SOURCES` to `clear;files,D:\archives,readwrite`

**Comment by strega-nil** [comment]
@ras0219 you _should_ be able to change the location of the default binary caches even without using this new `binarysources` thing.

**Comment by Neumann-A** [comment]
@strega-nil: do i just need to rerun clang-format to resolve the conflict? Or will you resolve the conflicts?

**Comment by strega-nil** [comment]
@Neumann-A yeah, it just needs to be LF and not CRLF.

**Comment by JackBoosY** [comment]
@strega-nil I think we need to improve the error message.

**Comment by strega-nil** [comment]
@JackBoosY yeah, I think you're right. I don't know how to get it to do what I want tho

**Comment by JackBoosY** [comment]
@strega-nil I'll do that. Don't worry.

**Comment by Neumann-A** [comment]
> So, the existing way to do exactly this is to define VCPKG_BINARY_SOURCES to clear;files,D:\archives,readwrite. Does this accomplish the desired behavior?

@ras0219-msft: Either I am doing something wrong or this deactivates binary caching completly

File: toolsrc/src/vcpkg/binarycaching.cpp [hunk 1]
```diff
@@ -663,6 +663,21 @@ namespace
     const ExpectedS<fs::path>& default_cache_path()
     {
         static auto cachepath = System::get_platform_cache_home().then([](fs::path p) -> ExpectedS<fs::path> {
+            auto maybe_cachepath = System::get_environment_variable("VCPKG_BINARY_CACHE");
+            if (auto p_str = maybe_cachepath.get())
+            {
+                const auto path = fs::u8path(*p_str); //(u8path is deprecated in C++20)
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/binarycaching.cpp:669]:
```suggestion
                const auto path = fs::u8path(*p_str);
```

`u8path` is deprecated in 20, but we won't be able to change over any time soon to `u8string`, and as of right now it's unimportant to change over. At the point where we can reasonably expect `char8_t` on all compilers, we'll actually start switching over.

File: toolsrc/src/vcpkg/binarycaching.cpp [hunk 2]
```diff
@@ -663,6 +663,21 @@ namespace
...
+                const auto status = fs::stdfs::status(path);
+                if (!fs::stdfs::exists(status))
+                    return {"Path to VCPKG_BINARY_CACHE does not exist: " + path.u8string(), expected_right_tag};
+                if (!fs::stdfs::is_directory(status))
+                    return {"Value of environment variable VCPKG_BINARY_CACHE is not a directory: " + path.u8string(),
+                            expected_right_tag};
+                if (!p.is_absolute())
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/binarycaching.cpp:676]:
```suggestion
                if (!path.is_absolute())
```

also this check should go first.

**Review by strega-nil — CHANGES_REQUESTED** [review]

**Review by strega-nil — APPROVED** [review]

**Comment by strega-nil** [comment]
@Neumann-A can you rename `VCPKG_BINARY_CACHE` to `VCPKG_DEFAULT_BINARY_CACHE`?

**Review by strega-nil — APPROVED** [review]

**Comment by strega-nil** [comment]
Thanks @Neumann-A !

**Pull request merged**